### ISSUE TITLE
Trying to run without setup.py test breaks tests.test_argparse under openSUSE

### ISSUE THREAD

**Issue by mcepl** [calmjs/calmjs#61]
This is not a duplicate of #58, because this is not about pytest. While packaging 3.4.2 for openSUSE and running the test suite by this command:
```
$ PYTHONPATH=/home/abuild/rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages
$ PYTHONDONTWRITEBYTECODE=1
$ python3.6 -m unittest discover -v -s /home/abuild/rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/
```
I get 560 tests passing, but one (and only one) test fails:
```
[   33s] ======================================================================
[   33s] ERROR: tests.test_argparse (unittest.loader._FailedTest)
[   33s] ----------------------------------------------------------------------
[   33s] ImportError: Failed to import test module: tests.test_argparse
[   33s] Traceback (most recent call last):
[   33s]   File "/usr/lib64/python3.6/unittest/loader.py", line 428, in _find_test_path
[   33s]     module = self._get_module_from_name(name)
[   33s]   File "/usr/lib64/python3.6/unittest/loader.py", line 369, in _get_module_from_name
[   33s]     __import__(name)
[   33s]   File "/home/abuild/rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/tests/test_argparse.py", line 9, in <module>
[   33s]     from calmjs.argparse import ArgumentParser
[   33s] ModuleNotFoundError: No module named 'calmjs.argparse'
[   33s]
[   33s]
[   33s] ----------------------------------------------------------------------
[   33s] Ran 561 tests in 28.186s
[   33s]
[   33s] FAILED (errors=1, skipped=1)
```
I don’t understand why only this one test fails and not others.

Complete [build log](https://github.com/calmjs/calmjs/files/7523679/_log.txt) with all versions of packages used and steps taken.

**Comment by metatoaster** [comment]
First off, `setup.py test` is no longer the standard - the current default way to invoke the test should be `python -m unittest` on a prepared test suite (as [documented](https://github.com/calmjs/calmjs#testing-the-installation), it should be `python -m unittest calmjs.tests.make_suite`).

When I attempted your method of test invocation (i.e. using `python -m unittest discover`), I was certainly able to reproduce `tests.test_argparse (unittest.loader._FailedTest) ... ERROR`, but also none of the other test modules can import any of the `calmjs` modules.

I don't know how your particular custom location install was able to import `calmjs.argparse` without issue for all the other tests while the `test_argparse` module cannot, as I am unable to replicate the issue, I am assuming `python -m unittest discover` has weird interactions with custom `PYTHONPATH`.

I can confirm that running `python -m unittest calmjs.tests.make_suite` instead will result in the correct test suite being created and tested.

**Comment by mcepl** [comment]
> I can confirm that running `python -m unittest calmjs.tests.make_suite` instead will result in the correct test suite being created and tested.

Are you certain?
```
[   35s] + PYTHONPATH=/home/abuild/rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages
[   35s] + PYTHONDONTWRITEBYTECODE=1
[   35s] + python3.6 -m unittest -v calmjs.tests.make_suite
[   35s] tests (unittest.loader._FailedTest) ... ERROR
[   35s]
[   35s] ======================================================================
[   35s] ERROR: tests (unittest.loader._FailedTest)
[   35s] ----------------------------------------------------------------------
[   35s] ImportError: Failed to import test module: tests
[   35s] Traceback (most recent call last):
[   35s]   File "/usr/lib64/python3.6/unittest/loader.py", line 153, in loadTestsFromName
[   35s]     module = __import__(module_name)
[   35s] ModuleNotFoundError: No module named 'calmjs.tests'
[   35s]
[   35s]
[   35s] ----------------------------------------------------------------------
[   35s] Ran 1 test in 0.000s
[   35s]
[   35s] FAILED (errors=1)
```
And again, [the complete build log](https://github.com/calmjs/calmjs/files/7524088/_log.txt).

And yes, of course I set up `$PYTHONPATH` to the location where the files are installed (here and in couple of thousand other Python packages where we run the test suite), because I believe it is better to test against the installed libraries rather than against the build tree.

**Comment by mcepl** [comment]
Just to make it clear, these is the tree with the installed files:
```
abuild@stitny:~> find rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/
abuild@stitny:~> find rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/argparse.py
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/artifact.py
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/base.py
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/calmjs.py
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/cli.py
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/command.py
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/dist.py
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/exc.py
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/indexer.py
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/interrogate.py
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/loaderplugin.py
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/module.py
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/npm.py
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/registry.py
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/runtime.py
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/toolchain.py
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/ui.py
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/utils.py
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/vlqsm.py
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/yarn.py
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/README.rst
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/testing
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/testing/__init__.py
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/testing/artifact.py
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/testing/mocks.py
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/testing/module.py
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/testing/spec.py
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/testing/utils.py
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/testing/module1
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/testing/module1/__init__.py
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/testing/module1/hello.js
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/testing/module1/__pycache__
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/testing/module1/__pycache__/__init__.cpython-36.opt-1.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/testing/module1/__pycache__/__init__.cpython-36.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/testing/module2
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/testing/module2/__init__.py
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/testing/module2/helper.js
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/testing/module2/index.js
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/testing/module2/mod
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/testing/module2/mod/helper.js
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/testing/module2/__pycache__
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/testing/module2/__pycache__/__init__.cpython-36.opt-1.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/testing/module2/__pycache__/__init__.cpython-36.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/testing/module3
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/testing/module3/__init__.py
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/testing/module3/module.py
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/testing/module3/runtime.py
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/testing/module3/math.js
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/testing/module3/mod
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/testing/module3/mod/index.js
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/testing/module3/__pycache__
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/testing/module3/__pycache__/__init__.cpython-36.opt-1.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/testing/module3/__pycache__/module.cpython-36.opt-1.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/testing/module3/__pycache__/runtime.cpython-36.opt-1.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/testing/module3/__pycache__/runtime.cpython-36.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/testing/module3/__pycache__/module.cpython-36.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/testing/module3/__pycache__/__init__.cpython-36.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/testing/module4
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/testing/module4/__init__.py
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/testing/module4/data.json
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/testing/module4/other.css
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/testing/module4/widget.js
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/testing/module4/widget.style
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/testing/module4/__pycache__
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/testing/module4/__pycache__/__init__.cpython-36.opt-1.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/testing/module4/__pycache__/__init__.cpython-36.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/testing/names
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/testing/names/typical.js
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/testing/names/unusual.js
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/testing/__pycache__
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/testing/__pycache__/mocks.cpython-36.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/testing/__pycache__/spec.cpython-36.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/testing/__pycache__/utils.cpython-36.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/testing/__pycache__/__init__.cpython-36.opt-1.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/testing/__pycache__/artifact.cpython-36.opt-1.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/testing/__pycache__/mocks.cpython-36.opt-1.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/testing/__pycache__/module.cpython-36.opt-1.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/testing/__pycache__/spec.cpython-36.opt-1.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/testing/__pycache__/utils.cpython-36.opt-1.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/testing/__pycache__/module.cpython-36.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/testing/__pycache__/artifact.cpython-36.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/testing/__pycache__/__init__.cpython-36.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/tests
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/tests/__init__.py
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/tests/test_argparse.py
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/tests/test_artifact.py
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/tests/test_base.py
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/tests/test_cli.py
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/tests/test_command.py
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/tests/test_dist.py
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/tests/test_indexer.py
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/tests/test_interrogate.py
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/tests/test_loaderplugin.py
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/tests/test_module.py
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/tests/test_registry.py
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/tests/test_runtime.py
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/tests/test_testing.py
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/tests/test_toolchain.py
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/tests/test_ui.py
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/tests/test_utils.py
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/tests/test_vlqsm.py
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/tests/__pycache__
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/tests/__pycache__/__init__.cpython-36.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/tests/__pycache__/test_argparse.cpython-36.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/tests/__pycache__/test_artifact.cpython-36.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/tests/__pycache__/test_base.cpython-36.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/tests/__pycache__/test_cli.cpython-36.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/tests/__pycache__/test_command.cpython-36.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/tests/__pycache__/test_dist.cpython-36.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/tests/__pycache__/test_indexer.cpython-36.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/tests/__pycache__/test_interrogate.cpython-36.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/tests/__pycache__/test_loaderplugin.cpython-36.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/tests/__pycache__/test_module.cpython-36.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/tests/__pycache__/test_registry.cpython-36.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/tests/__pycache__/test_runtime.cpython-36.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/tests/__pycache__/test_testing.cpython-36.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/tests/__pycache__/test_toolchain.cpython-36.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/tests/__pycache__/test_ui.cpython-36.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/tests/__pycache__/test_utils.cpython-36.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/tests/__pycache__/__init__.cpython-36.opt-1.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/tests/__pycache__/test_argparse.cpython-36.opt-1.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/tests/__pycache__/test_artifact.cpython-36.opt-1.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/tests/__pycache__/test_base.cpython-36.opt-1.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/tests/__pycache__/test_cli.cpython-36.opt-1.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/tests/__pycache__/test_command.cpython-36.opt-1.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/tests/__pycache__/test_dist.cpython-36.opt-1.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/tests/__pycache__/test_indexer.cpython-36.opt-1.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/tests/__pycache__/test_interrogate.cpython-36.opt-1.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/tests/__pycache__/test_loaderplugin.cpython-36.opt-1.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/tests/__pycache__/test_module.cpython-36.opt-1.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/tests/__pycache__/test_registry.cpython-36.opt-1.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/tests/__pycache__/test_runtime.cpython-36.opt-1.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/tests/__pycache__/test_testing.cpython-36.opt-1.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/tests/__pycache__/test_toolchain.cpython-36.opt-1.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/tests/__pycache__/test_ui.cpython-36.opt-1.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/tests/__pycache__/test_utils.cpython-36.opt-1.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/tests/__pycache__/test_vlqsm.cpython-36.opt-1.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/tests/__pycache__/test_vlqsm.cpython-36.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/__pycache__
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/__pycache__/argparse.cpython-36.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/__pycache__/artifact.cpython-36.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/__pycache__/base.cpython-36.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/__pycache__/cli.cpython-36.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/__pycache__/command.cpython-36.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/__pycache__/dist.cpython-36.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/__pycache__/indexer.cpython-36.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/__pycache__/interrogate.cpython-36.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/__pycache__/loaderplugin.cpython-36.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/__pycache__/module.cpython-36.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/__pycache__/npm.cpython-36.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/__pycache__/registry.cpython-36.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/__pycache__/runtime.cpython-36.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/__pycache__/toolchain.cpython-36.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/__pycache__/ui.cpython-36.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/__pycache__/utils.cpython-36.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/__pycache__/vlqsm.cpython-36.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/__pycache__/yarn.cpython-36.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/__pycache__/argparse.cpython-36.opt-1.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/__pycache__/artifact.cpython-36.opt-1.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/__pycache__/base.cpython-36.opt-1.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/__pycache__/calmjs.cpython-36.opt-1.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/__pycache__/cli.cpython-36.opt-1.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/__pycache__/command.cpython-36.opt-1.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/__pycache__/dist.cpython-36.opt-1.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/__pycache__/exc.cpython-36.opt-1.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/__pycache__/indexer.cpython-36.opt-1.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/__pycache__/interrogate.cpython-36.opt-1.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/__pycache__/loaderplugin.cpython-36.opt-1.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/__pycache__/module.cpython-36.opt-1.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/__pycache__/npm.cpython-36.opt-1.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/__pycache__/registry.cpython-36.opt-1.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/__pycache__/runtime.cpython-36.opt-1.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/__pycache__/toolchain.cpython-36.opt-1.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/__pycache__/ui.cpython-36.opt-1.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/__pycache__/utils.cpython-36.opt-1.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/__pycache__/vlqsm.cpython-36.opt-1.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/__pycache__/yarn.cpython-36.opt-1.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/__pycache__/exc.cpython-36.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/__pycache__/calmjs.cpython-36.pyc
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs/__init__.py
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs-3.4.2-py3.6.egg-info
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs-3.4.2-py3.6.egg-info/PKG-INFO
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs-3.4.2-py3.6.egg-info/dependency_links.txt
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs-3.4.2-py3.6.egg-info/entry_points.txt
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs-3.4.2-py3.6.egg-info/namespace_packages.txt
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs-3.4.2-py3.6.egg-info/requires.txt
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs-3.4.2-py3.6.egg-info/SOURCES.txt
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs-3.4.2-py3.6.egg-info/top_level.txt
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs-3.4.2-py3.6.egg-info/not-zip-safe
rpmbuild/BUILDROOT/python-calmjs-3.4.2-0.x86_64/usr/lib/python3.6/site-packages/calmjs-3.4.2-py3.6-nspkg.pth
abuild@stitny:~>
```

**Comment by metatoaster** [comment]
> And yes, of course I set up `$PYTHONPATH` to the location where the files are installed (here and in couple of thousand other Python packages where we run the test suite), because I believe it is better to test against the installed libraries rather than against the build tree.

Of course, and the standard `python -m unittest` will test against whatever is available and visible to Python.

It also turns out discovery works again, so I am not sure what Python was doing initially.  Attached is a log of my attempt to replicate the issue you are seeing.

[log.txt](https://github.com/calmjs/calmjs/files/7524135/log.txt)

**Comment by metatoaster** [comment]
Looking at the build script in the linked [commit](https://github.com/bmwiedemann/openSUSE/commit/a14788a2e07ee0f392d808455acb8243e4f70a9f), I see this line:

```
%{python_expand touch %{buildroot}%{$python_sitelib}/calmjs/__init__.py
```

Executing that line (i.e. creating the `__init__.py` file) in my context reproduced the error with `calmjs.tests` not being found:

```console
$ touch /tmp/python-calmjs-3.4.2/usr/lib/python3.9/site-packages/calmjs/__init__.py
$ PYTHONPATH=/tmp/python-calmjs-3.4.2/usr/lib/python3.9/site-packages python3.9 -m unittest calmjs.tests.make_suite
E
======================================================================
ERROR: tests (unittest.loader._FailedTest)
----------------------------------------------------------------------
ImportError: Failed to import test module: tests
Traceback (most recent call last):
  File "/usr/lib/python3.9/unittest/loader.py", line 154, in loadTestsFromName
    module = __import__(module_name)
ModuleNotFoundError: No module named 'calmjs.tests'


----------------------------------------------------------------------
Ran 1 test in 0.000s
```

Moreover, it also perfectly reproduce the import error initially reported, due to subsequent imports being correctly initialized (despite failing first):

```console
$ PYTHONPATH=/tmp/python-calmjs-3.4.2/usr/lib/python3.9/site-packages python3.9 -m unittest discover -s /tmp/python-calmjs-3.4.2/usr/lib/python3.9/site-packages/calmjs/tests/
/tmp/python-calmjs-3.4.2/usr/lib/python3.9/site-packages/calmjs/cli.py:234: RuntimeWarning: Unable to locate the 'yarn' binary or runtime; default module level functions will not work. Please either provide PATH and/or update PATH environment variable with one that provides 'yarn'; or specify a working NODE_PATH environment variable with yarn installed; or have install 'yarn' into the current working directory (/home/toaster/work) either through npm or calmjs framework for this package. Restart or reload this module once that is done. Alternatively, create a manual Driver instance for 'yarn' with explicitly defined arguments.
  warnings.warn(msg, RuntimeWarning)
encountered unexpected error while trying to find version of 'None':
Traceback (most recent call last):
  File "/tmp/python-calmjs-3.4.2/usr/lib/python3.9/site-packages/calmjs/cli.py", line 52, in get_bin_version_str
    prog = _get_exec_binary(bin_path, kw)
  File "/tmp/python-calmjs-3.4.2/usr/lib/python3.9/site-packages/calmjs/base.py", line 75, in _get_exec_binary
    binary = which(binary, path=kw.get('env', {}).get('PATH'))
  File "/tmp/python-calmjs-3.4.2/usr/lib/python3.9/site-packages/calmjs/utils.py", line 135, in which
    if os.path.dirname(cmd):
  File "/usr/lib/python3.9/posixpath.py", line 152, in dirname
    p = os.fspath(p)
TypeError: expected str, bytes or os.PathLike object, not NoneType
E..............................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................ss.ss...s.................sss.s
======================================================================
ERROR: test_argparse (unittest.loader._FailedTest)
----------------------------------------------------------------------
ImportError: Failed to import test module: test_argparse
Traceback (most recent call last):
  File "/usr/lib/python3.9/unittest/loader.py", line 436, in _find_test_path
    module = self._get_module_from_name(name)
  File "/usr/lib/python3.9/unittest/loader.py", line 377, in _get_module_from_name
    __import__(name)
  File "/tmp/python-calmjs-3.4.2/usr/lib/python3.9/site-packages/calmjs/tests/test_argparse.py", line 9, in <module>
    from calmjs.argparse import ArgumentParser
ModuleNotFoundError: No module named 'calmjs.argparse'


----------------------------------------------------------------------
Ran 638 tests in 4.264s

FAILED (errors=1, skipped=9)
```

Removing the offending `__init__.py` corrects the issue:

```console
$ rm /tmp/python-calmjs-3.4.2/usr/lib/python3.9/site-packages/calmjs/__init__.py
rm: remove regular empty file '/tmp/python-calmjs-3.4.2/usr/lib/python3.9/site-packages/calmjs/__init__.py'? y
$ PYTHONPATH=/tmp/python-calmjs-3.4.2/usr/lib/python3.9/site-packages python3.9 -m unittest discover -s /tmp/python-calmjs-3.4.2/usr/lib/python3.9/site-packages/calmjs/tests/
/tmp/python-calmjs-3.4.2/usr/lib/python3.9/site-packages/calmjs/cli.py:234: RuntimeWarning: Unable to locate the 'yarn' binary or runtime; default module level functions will not work. Please either provide PATH and/or update PATH environment variable with one that provides 'yarn'; or specify a working NODE_PATH environment variable with yarn installed; or have install 'yarn' into the current working directory (/home/toaster/work) either through npm or calmjs framework for this package. Restart or reload this module once that is done. Alternatively, create a manual Driver instance for 'yarn' with explicitly defined arguments.
  warnings.warn(msg, RuntimeWarning)
encountered unexpected error while trying to find version of 'None':
Traceback (most recent call last):
  File "/tmp/python-calmjs-3.4.2/usr/lib/python3.9/site-packages/calmjs/cli.py", line 52, in get_bin_version_str
    prog = _get_exec_binary(bin_path, kw)
  File "/tmp/python-calmjs-3.4.2/usr/lib/python3.9/site-packages/calmjs/base.py", line 75, in _get_exec_binary
    binary = which(binary, path=kw.get('env', {}).get('PATH'))
  File "/tmp/python-calmjs-3.4.2/usr/lib/python3.9/site-packages/calmjs/utils.py", line 135, in which
    if os.path.dirname(cmd):
  File "/usr/lib/python3.9/posixpath.py", line 152, in dirname
    p = os.fspath(p)
TypeError: expected str, bytes or os.PathLike object, not NoneType
.............................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................ss.ss...s.................sss.s
----------------------------------------------------------------------
Ran 668 tests in 4.585s

OK (skipped=9)
```

Removing that new files restores the expected behavior.  Do not create `__init__.py` files as `calmjs` is declared as a namespace (as [documented in this section](https://github.com/calmjs/calmjs#random-importerror-when-trying-to-import-from-the-calmjs-namespace), it covers the symptoms seen) and having that `__init__.py` file breaks that namespace functionality.

Again, this test failure is only a symptom of a broken namespace package, which was a present bug in old versions of `setuptools` where it incorrectly configured a `__init__.py` hence a full reinstallation was required to ensure that this `__init__.py` file do not exist for Python 3 installations.  Users of the openSUSE `calmjs` package will experience this random import breakage if this is extraneous `__init__.py` file is not removed.

**Comment by mcepl** [comment]
Thank you very much! Yes, non-creation of that \_\_init\_\_.py file helped. Now the test suite passes only with these warnings:
```
[   91s] test_transpiler_sourcemap (test_toolchain.ES5ToolchainTestCase) ... WARNING: Couldn't write lextab module 'calmjs.parse.parsers.lextab_es5_py3_ply3_11'. [Errno 13] Permission denied: '/usr/lib/python3.8/site-packages/calmjs/parse/parsers/lextab_es5_py3_ply3_11.py'
[   91s] WARNING: Token 'BLOCK_COMMENT' defined, but not used
[   91s] WARNING: Token 'LINE_COMMENT' defined, but not used
[   91s] WARNING: Token 'LINE_TERMINATOR' defined, but not used
[   91s] WARNING: There are 3 unused tokens
[   91s] WARNING: Couldn't create 'calmjs.parse.parsers.yacctab_es5_py3_ply3_11'. [Errno 13] Permission denied: '/usr/lib/python3.8/site-packages/calmjs/parse/parsers/yacctab_es5_py3_ply3_11.py'
[   91s] ok
```
Any idea what’s going on, please?

Thank you very much again.

**Comment by metatoaster** [comment]
That's due to the `calmjs.parse` being packaged incorrectly on openSUSE, please refer to [manual optimization](https://github.com/calmjs/calmjs.parse#manual-optimization) to create the missing generated modules for the meantime (`ply` will attempt to generate them, but can't write them out because the current user has no permission).

A more permanent remedy will involve openSUSE making a new release of `calmjs.parse` using the 1.3.0 source tarball, which includes that generated module that is not under version control; refer to calmjs/calmjs.parse#34 for the discussion that lead to this outcome.

**Comment by mcepl** [comment]
That manual optimization should go probably to the installation phase, right? Like in https://build.opensuse.org/request/show/931802 ?

**Comment by metatoaster** [comment]
Correct, installation phase will work although something something similar to what [ArchLinux](https://github.com/archlinux/svntogit-community/blob/9f1d05d528280adf862d65470c711c513b340b95/trunk/PKGBUILD#L18-L25) does (i.e. run `PYTHONPATH=src python -m calmjs.parse.parsers.optimize` before installation from the Github tag tarball root directory) should also work.

**Comment by mcepl** [comment]
> Correct, installation phase will work although something something similar to what [ArchLinux](https://github.com/archlinux/svntogit-community/blob/9f1d05d528280adf862d65470c711c513b340b95/trunk/PKGBUILD#L18-L25) does (i.e. run `PYTHONPATH=src python -m calmjs.parse.parsers.optimize` before installation from the Github tag tarball root directory) should also work.

Right, we are contrary to ArchLinux build for multiple versions of Python at once (currently 3.6, 3.8, and 3.9), so the installation (or build) phase is better, the preparation phase is universal for all versions together.

**Comment by metatoaster** [comment]
Alright, just checked the [latest openSUSE build for `calmjs.parse`](https://build.opensuse.org/request/show/931812) and I think that should do the trick.  It would be useful if you can confirm that the warnings involving the lextab module and the permission denied warnings are squelched after the fix, and please close the issue if this packaging issue is resolved for openSUSE.

**Comment by mcepl** [comment]
Well, I am not sure what I see in the [complete build log](https://github.com/calmjs/calmjs/files/7552376/_log.txt). During the ``python3 setup.py install`` (63s) I get this error:
```
[   63s] running install_scripts
[   63s] /usr/bin/python3.8: Error while finding module specification for 'calmjs.parse.parsers.optimize' (ModuleNotFoundError: No module named 'calmjs')
```
and when I run ``calmjs.parse.parsers.optimize`` (63-67s) it seems to work fine:
```
[   63s] + PYTHONPATH=/home/abuild/rpmbuild/BUILDROOT/python-calmjs.parse-1.3.0-12.1.aarch64/usr/lib/python3.6/site-packages
[   63s] + python3.6 -m calmjs.parse.parsers.optimize
[   67s] WARNING: Token 'BLOCK_COMMENT' defined, but not used
[   67s] WARNING: Token 'LINE_COMMENT' defined, but not used
[   67s] WARNING: Token 'LINE_TERMINATOR' defined, but not used
[   67s] WARNING: There are 3 unused tokens
```
(the same for Python 3.8 and Python 3.9 later in the log).

Is this correct?

**Comment by metatoaster** [comment]
> Well, I am not sure what I see in the [complete build log](https://github.com/calmjs/calmjs/files/7552376/_log.txt). During the `python3 setup.py install` (63s) I get this error:

That should be a spurious message caused by the [`InstallHook` trying to call that same line and failing](https://github.com/calmjs/calmjs.parse/blob/1.3.0/setup.py#L8-L13) (I really should remove that hook or investigate in another method to set that up, as that step was initially created as an attempt to fix this very issue, though given that failure will result in nothing bad happening I left it there), though the [subsequently manually invoked optimize step in the spec file](https://github.com/bmwiedemann/openSUSE/blob/aaec684ec/packages/p/python-calmjs.parse/python-calmjs.parse.spec#L49) should then generate that correctly, generating those unused token warnings and as long as the following files are generated and be made part of set of installed files

```
calmjs/parse/parsers/lextab_es5_py3_ply3_11.py
calmjs/parse/parsers/yacctab_es5_py3_ply3_11.py
```

Provided those generated files are installed it should be fine, though I can't help but note the absence of their `pyc` in the log file.  I think pre-generating the lextab/yacctab files in the build step (using a `PYTHONPATH` as mentioned earlier) may be better in that ensuring those files are available before the `%python_build` step is run.

**Comment by mcepl** [comment]
> Though I can't help but note the absence of their `pyc` in the log file you have provided. I think pre-generating the lextab/yacctab files in the build step (using a `PYTHONPATH` as mentioned earlier) may be better in that ensuring those files are available before the `%python_build` step is run.

Thanks, will do.

**Comment by mcepl** [comment]
OK, so
```spec
 %build
 export LANG=en_US.UTF-8
 %python_build
%{python_expand PYTHONPATH=build/lib $python -m calmjs.parse.parsers.optimize
$python -m compileall -j 0 -f build/
}
```
seems to work, which however brings back the wisdom of ArchLinux and doing the optimization in `%prep` stage. How much are these tables independent of the Python version?

**Comment by metatoaster** [comment]
> How much are these tables independent of the Python version?

They are compatible across major versions, i.e. Ones produced in Python 3 will not work in Python 2, but not necessarily vice-versa. My official [releases on PyPI](https://pypi.org/project/calmjs.parse/#files) as wheels only break them down as far as major Python versions.

**Comment by mcepl** [comment]
OK, I will rather be safe than sorry and regenerate them for each Python version. And that means this ticket can be closed. Thank you very much for all your kind help!

**Comment by metatoaster** [comment]
You're welcome! Thank you for contacting me about the packaging issues on openSUSE, glad to have helped to make this work.
